applybn/feature_selection/bn_nmi_feature_selector.py

```diff
@@ -0,0 +1,166 @@
+      x_min, x_max = arr_x.min(), arr_x.max()
+      y_min, y_max = arr_y.min(), arr_y.max()
+      
+      hist, _, _ = np.histogram2d(
+        arr_x, arr_y,
+        bins=[x_max - x_min + 1, y_max - y_min + 1],
+        range=[[x_min, x_max + 1], [y_min, y_max + 1]]
+      )
+      return hist.flatten().astype(int).tolist()
+
+
+  def _normalized_mutual_information(self, a, b):
+    '''
+    Calculate Normalized Mutual Information (NMI) for pair of features
+
+    Args: 
+        a, b - pair of pd.Series to calculate NMI.
+    '''
+    if (a.name, None) not in self.entropy_map:
+      self.entropy_map[(a.name, None)] = float(drv.entropy(self._discreticise(a)))
+    if (b.name, None) not in self.entropy_map:
+      self.entropy_map[(b.name, None)] = float(drv.entropy(self._discreticise(b)))
+    if (a.name, b.name) not in self.entropy_map:
+      self.entropy_map[(a.name, b.name)] = float(drv.entropy_joint(self._discreticise(a,b)))
+
+    Ha = self.entropy_map[(a.name, None)]
+    Hb = self.entropy_map[(b.name, None)]
+    Hab = self.entropy_map[(a.name, b.name)]
+    return (Ha + Hb - Hab) / (2 * max(Ha, Hb))
+
+  def _prepare_dataset(self, df: pd.DataFrame):
+    '''
+    Prepare the dataset, discretize the fields, remove rows
+    Return a ready-to-use dataset
+    
+    Args: 
+        df - data frame to be cleaned.
+    '''
+    clean_df = pd.DataFrame()
+    for col in df:
+      if pd.api.types.is_integer_dtype(df[col].dtype):
+        # integers can be discreticised this way, without binning
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+      elif pd.api.types.is_numeric_dtype(df[col].dtype):
+        # discretisize to bin_count bins
+        clean_df[col] = pd.cut(df[col], bins=self.bin_count, labels=self.bin_labels, duplicates='drop')
+        clean_df[col] = clean_df[col].cat.add_categories(self.bin_count + 1).fillna(self.bin_count + 1)
+      else:
+        # pandas cut does not work with string features, this is done manually
+        # it is assumed that the user has removed the ID-like features, otherwise there will be a memory error later
+        df_map = { l:i for i,l in enumerate(df[col].unique())}
+        clean_df[col] = df[col].map(df_map)
+    return clean_df
+
+  def fit(self, X: pd.DataFrame, y: pd.Series = None):    
+    if y is None:
+      return self
+
+    df = X
+    df[y.name] = y
+    target = y.name
+    self.initial_features = list(df.columns)
+
+    clean_df = self._prepare_dataset(df)
+    if self.verboose:
+      logging.info('Dataset is ready')
+    nmi_map = dict()
+    self.entropy_map = dict()
+    pre_features = []
+    for col in clean_df:
+      if col == target:
+        continue
+      if (col, target) not in nmi_map.keys():
+        nmi_map[(col, target)] = self._normalized_mutual_information(clean_df[col], clean_df[target])
+      nmi = nmi_map[(col, target)]
+      if nmi > self.threshhold:
+        pre_features.append(col)
```

Comment:
Here, I assume, that various pandas methods could be used to avoid running pandas inside loop the same as in `_prepare_dataset` method.